### PR TITLE
Update database for chat module

### PR DESCRIPTION
**Pull request by Frank-Skywalker**

### CODE REVIEW

File: pkg/chat/model.go [hunk 1]
```diff
@@ -0,0 +1,37 @@
+package chat
+
+import "time"
+
+type Chat struct {
```
Comment (AsterNighT) [inline pkg/chat/model.go:3]: Pls make it a gorm model

Comment (AsterNighT) [reply pkg/chat/model.go:3]: And use association rather than specify ID

File: pkg/chat/model.go [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+
+type Message struct{
+	ID			uint
+	ChatID		uint
+	Timestamp	Time
+	Type		uint				//Texts, pictures....
+	Content		string
+	PictureURL  string
```
Comment (AsterNighT) [inline pkg/chat/model.go]: `PictureURL` can be part of content. You can make content an escaped string for that.

File: pkg/chat/model.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+type Message struct{
+	ID			uint
+	ChatID		uint
+	Timestamp	Time
+	Type		uint				//Texts, pictures....
+	Content		string
+	PictureURL  string 
+
+	//primary key: ID&chatID
```
Comment (AsterNighT) [inline pkg/chat/model.go]: And primary key should be ID only.

File: pkg/chat/model.go [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+
+//Question catagories
+//For standard query,each catagory correspond with a set of fixed questions
+type Catagory struct{
+	ID			uint
+	Name		string
+	Keywords	[]string
```
Comment (AsterNighT) [inline pkg/chat/model.go]: I don't quite understand what is `Keywords` and `Questions`, can you make some explainations on those?

**Review by AsterNighT — CHANGES_REQUESTED** [review]
PTAL

File: pkg/chat/model.go [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+	Chat      Chat //Default foreign key:ID in Chat
+	CreatedAt int
+	Type      uint   //Texts, pictures....
+	Content   string //Use escaped string for PictureURL
+}
+
+/*
+Question catagories
+-------------------------------------------------------------------
+If the patient's message contains some keyword specified in DB,
```
Comment (AsterNighT) [inline pkg/chat/model.go:21]: Umm, if I do not misunderstand, you search the Keywords for catagories, and get questions through catagory, am I right on this?

Comment (Frank-Skywalker) [reply pkg/chat/model.go:21]: Yes!!!

Comment (AsterNighT) [reply pkg/chat/model.go:21]: That way would it be better if you put the relationship as: A many to many association from key to catagory, and a catagory has many questions

Comment (Frank-Skywalker) [reply pkg/chat/model.go:21]: copy that

File: pkg/chat/model.go [hunk 6]
```diff
@@ -25,6 +25,11 @@ Doctor's work can be reduced by automatically showing right questions.
 type Catagory struct {
 	ID        uint `gorm:"primaryKey"`
 	Name      string
-	Keywords  *string
-	Questions *string //Each catagory correspond with a set of fixed questions
+	Keywords  []Keyword `gorm:"many2many:catagory_keywords;"`
```
Comment (AsterNighT) [inline pkg/chat/model.go]: You'll need the reversed direction. You get catagory by keywords, not get keywords by catagory

**Review by AsterNighT — CHANGES_REQUESTED** [review]

File: pkg/router/router.go [hunk 7]
```diff
@@ -4,7 +4,7 @@ import (
 	"net/http"
 
 	"github.com/AsterNighT/software-engineering-backend/api"
-	_ "github.com/AsterNighT/software-engineering-backend/docs" // swagger doc
+	//	_ "github.com/AsterNighT/software-engineering-backend/docs" // swagger doc
```
Comment (AsterNighT) [inline pkg/router/router.go]: Pls do not comment out this, keep docs in your folder and ignore it rather than delete it

**Review by AsterNighT — CHANGES_REQUESTED** [review]

File: pkg/router/router.go [hunk 8]
```diff
@@ -3,8 +3,7 @@ package router
 import (
 	"net/http"
 
-	"github.com/AsterNighT/software-engineering-backend/api"
-	//	_ "github.com/AsterNighT/software-engineering-backend/docs" // swagger doc
```
Comment (AsterNighT) [inline pkg/router/router.go]: No you should not delete `_ "github.com/AsterNighT/software-engineering-backend/docs" // swagger doc`, this is intended.

**Review by AsterNighT — CHANGES_REQUESTED** [review]

File: go.mod [hunk 9]
```diff
@@ -4,13 +4,40 @@ go 1.16
 
 require (
 	github.com/alecthomas/template v0.0.0-20190718012654-fb15b899a751
+	github.com/armon/consul-api v0.0.0-20180202201655-eb2c6b5be1b6 // indirect
```
Comment (AsterNighT) [inline go.mod]: It seems that no new packages are used in the api, where do these new requirements come from?

File: pkg/chat/handler.go [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+
+type ChatHandler struct {
+}
+
+// @Summary New a chat
+// @Description
+// @Tags Chat
+// @Produce json
+// @Param chatDetail body Chat true "patient ID, doctor ID and other chat details"
+// @Success 200 {object} api.ReturnedData{data=Chat}
```
Comment (AsterNighT) [inline pkg/chat/handler.go]:
```suggestion
// @Success 200 {object} api.ReturnedData{}
```

File: pkg/chat/handler.go [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+// @Router /patient/{patientID}/chat [POST]
+func (h *ChatHandler) NewChat(c echo.Context) error {
+	// ...
+	c.Logger().Debug("hello world")
+	return c.JSON(200, api.Return("NewChat", nil))
```
Comment (AsterNighT) [inline pkg/chat/handler.go]:
```suggestion
	return c.JSON(200, api.Return("ok", nil))
```

File: pkg/chat/handler.go [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+}
+
+// @Summary Delete a chat
+// @Description
+// @Tags Chat
+// @Produce json
+// @Param chatID path uint true "chat ID"
```
Comment (AsterNighT) [inline pkg/chat/handler.go:35]: And patientID, check the rest of the code for similar issues.

File: docs/docs.go [hunk 13]
```diff
@@ -2,3 +2,537 @@
 // This file was generated by swaggo/swag
 
 package docs
+
```
Comment (AsterNighT) [inline docs/docs.go]: get the `.gitignore` and `placeholder.go` from master and do not upload new docs.go anymore. My bad on this,

**Review by AsterNighT — CHANGES_REQUESTED** [review]

File: go.mod (hunk 9 above)
Comment (Frank-Skywalker) [reply go.mod:7]: Emmm...maybe i'll should use go.mod in previous version...

Comment (Frank-Skywalker) [reply go.mod]: I'll try using go.mod in previous version.

File: docs/docs.go (hunk 13 above)
Comment (Frank-Skywalker) [reply docs/docs.go]: Emmm...did you mean delete the docs.go file?

File: pkg/chat/handler.go [hunk 14]
```diff
@@ -13,114 +13,124 @@ type ChatHandler struct {
...
 
 // @Summary Get a message by message ID
 // @Description
 // @Tags Chat
 // @Produce json
+// @Param patientID path uint true "patient ID"
+// @Param chatID path uint true "chat ID"
 // @Param messageID path uint true "message ID"
-// @Success 200 {array} api.ReturnedData{data=Message}
+// @Success 200 {array} api.ReturnedData{}
```
Comment (AsterNighT) [inline pkg/chat/handler.go]:
```suggestion
// @Success 200 {object} api.ReturnedData{data=Message}
```

File: pkg/chat/handler.go [hunk 15]
```diff
@@ -13,114 +13,124 @@ type ChatHandler struct {
...
 // @Success 200 {object} api.ReturnedData{}
 // @Router /patient/{patientID}/chat/{chatID}/message/{messageID} [DELETE]
 func (h *ChatHandler) DeleteMessage(c echo.Context) error {
 	// ...
 	c.Logger().Debug("hello world")
-	return c.JSON(200, api.Return("DeleteMessage", nil))
+	return c.JSON(200, api.Return("ok", nil))
 }
 
 // @Summary Get a message ID list by chat ID
```
Comment (AsterNighT) [inline pkg/chat/handler.go]: Don't return IDs, return data directly.
What can we do with the IDs?

File: pkg/chat/handler.go [hunk 16]
```diff
@@ -13,114 +13,124 @@ type ChatHandler struct {
...
 // @Description
 // @Tags Chat
 // @Produce json
+// @Param patientID path uint true "patient ID"
 // @Param chatID path uint true "chat ID"
-// @Success 200 {array} api.ReturnedData{data=[]uint}
+// @Success 200 {array} api.ReturnedData{}
```
Comment (AsterNighT) [inline pkg/chat/handler.go]: Ditto

File: pkg/chat/handler.go [hunk 17]
```diff
@@ -13,114 +13,124 @@ type ChatHandler struct {
...
 }
 
 // @Summary Get questions by catagory id
 // @Description
 // @Tags Chat
 // @Produce json
+// @Param keywordID path uint true "keyword ID"
 // @Param catagoryID path uint true "catagory ID"
-// @Success 200 {object} api.ReturnedData{data=[]string}
+// @Success 200 {object} api.ReturnedData{}
```
Comment (AsterNighT) [inline pkg/chat/handler.go]: Ditto

File: pkg/chat/handler.go [hunk 18]
```diff
@@ -13,114 +13,124 @@ type ChatHandler struct {
...
+	return c.JSON(200, api.Return("ok", nil))
 }
 
 // @Summary Get catagorys by keyword id
 // @Description
 // @Tags Chat
 // @Produce json
 // @Param keywordID path uint true "keyword ID"
-// @Success 200 {object} api.ReturnedData{data=[]uint}
+// @Success 200 {object} api.ReturnedData{}
```
Comment (AsterNighT) [inline pkg/chat/handler.go]: Ditto

File: pkg/chat/handler.go [hunk 19]
```diff
@@ -13,114 +13,124 @@ type ChatHandler struct {
...
 	c.Logger().Debug("hello world")
-	return c.JSON(200, api.Return("GetCatagorysByKeywordID", nil))
+	return c.JSON(200, api.Return("ok", nil))
 }
 
 // @Summary Get questions by catagory id
 // @Description
 // @Tags Chat
 // @Produce json
+// @Param keywordID path uint true "keyword ID"
```
Comment (AsterNighT) [inline pkg/chat/handler.go]: Why do we need keywordID here?

**Review by AsterNighT — CHANGES_REQUESTED** [review]

File: docs/docs.go (hunk 13 above)
Comment (AsterNighT) [reply docs/docs.go]: Yes

File: pkg/chat/handler.go (hunk 10 above)
Comment (Frank-Skywalker) [reply pkg/chat/handler.go]: I don't understand it...Why do we need to return data(in Message form) when a message is newed,but we don't need to return data (in Chat form) when a chat is newed?

**Comment by Frank-Skywalker** [comment]
I don't understand it...Why do we need to return data(in Message form) when a message is newed,but we don't need to return data (in Chat form) when a chat is newed?

**Comment by Frank-Skywalker** [comment]
Should we implement this module using client-server model or peer-to-peer model?

**Comment by AsterNighT** [comment]
> I don't understand it...Why do we need to return data(in Message form) when a message is newed,but we don't need to return data (in Chat form) when a chat is newed?

We need to return data in neither of the two situation. I may miss some of the parts.

**Comment by AsterNighT** [comment]
> Should we implement this module using client-server model or peer-to-peer model?

C-S would be preferable.

**Review by AsterNighT — COMMENTED** [review]
LGTM now, but pls revert your change in go.sum

File: pkg/chat/handler.go [hunk 20]
```diff
@@ -128,7 +144,7 @@ func (h *ChatHandler) GetCatagorysByMessageID(c echo.Context) error {
 // @Param catagoryID path uint true "catagory ID"
 // @Success 200 {object} api.ReturnedData{data=[]string}
 // @Router /keyword/{keywordID}/catagory/{catagoryID}  [GET]
-func (h *ChatHandler) GetQuestiosByCatagoryID(c echo.Context) error {
+func (h *KeywordHandler) GetQuestionsByCatagoryID(c echo.Context) error {
```
Comment (AsterNighT) [inline pkg/chat/handler.go]: Why do we need keywordID here?

Comment (Frank-Skywalker) [reply pkg/chat/handler.go]: Emmm...do you mean that the router is wrong or the keywordhandler is unnecessary?

Comment (AsterNighT) [reply pkg/chat/handler.go]: Well that's ok

**Comment by AsterNighT** [comment]
LGTM now, please resolve the conflict and I'll merge this

**Pull request merged**